Documentation page improvements

freemansw1 opened the pull request:
This PR brings over most of the documentation page improvements from #127 , and should pair nicely with #138 . 

* [x] Have you followed our guidelines in CONTRIBUTING.md? 
* [x] Have you self-reviewed your code and corrected any misspellings? 
* [x] Have you written documentation that is easy to understand?
* [x] Have you written descriptive commit messages? 
* [ ] Have you added NumPy docstrings for newly added functions? 
* [ ] Have you formatted your code using black? 
* [ ] If you have introduced a new functionality, have you added adequate unit tests?
* [ ] Have all tests passed in your local clone? 
* [ ] If you have introduced a new functionality, have you added an example notebook?
* [x] Have you kept your pull request small and limited so that it is easy to review? 
* [x] Have the newest changes from this branch been merged? 

To get his feet wet on reviewing, I've requested @pjmarinescu as a reviewer here. As one of the original tobac devs, I think his perspective on this will be really useful.

freemansw1 commented:
Ah- I've now added readthedocs building to PRs, which has just failed! It builds locally, but clearly, there is an issue with readthedocs. I'll work on revising the PR shortly.

freemansw1 commented:
Thanks to readthedocs, you can preview these new documents here: https://tobac--150.org.readthedocs.build/en/150/ (or by hitting `details` on the readthedocs action. 

Also, given the relation to #138 , I've requested @snilsn as a third reviewer, if he is interested and has time.

freemansw1 commented:
I've gone through and made some substantial changes to this PR, including a new page describing some of the feature detection parameters and three new Jupyter notebooks addressing the same. I do plan to expand on this and do similar work on both the segmentation section and tracking section.

freemansw1 commented:
I wonder if we should move/copy the examples from the examples folder over to the readthedocs? Maybe that's a separate issue...

snilsn commented:
I really like the new notebooks. Together with an updated documentation this should really make getting started with tobac easier.

We just have to keep in mind, that @fsenf and I have been working on a [very similar thing ](https://github.com/tobac-project/tobac-tutorials/tree/master/tutorials/Basics) in _tobac-tutorials_, based on v2. I'm planning to make a PR with notebooks on tracking and segmentation there soon.

I think more examples can never hurt, but we should probably discuss how to proceed to avoid redundancy.

freemansw1 commented:
Yes, agreed that I want to avoid duplication. I hadn't seen your new `basics` folder, and some of this definitely overlaps, although my code is written for v1.x rather than v2.x. As we progress into v2.x (and/or #143 is resolved), I think we should bring the basics notebooks over to the readthedocs to replace these notebooks? What are your thoughts? 

Regardless, I think that we should discuss this at our developers' meeting this week.

pjmarinescu commented on doc/installation.rst:
```diff
@@ -1,42 +1,38 @@
-You can also install conda via pip, which is mainly interesting for development purposed or to use specific development branches for the Github repository.
+You can also install conda via pip, which is mainly interesting for development purposes or using specific development branches for the Github repository.
 
 The follwoing python packages are required (including dependencies of these packages):
```
Typo here "following"

pjmarinescu commented on doc/threshold_detection_parameters.rst:
```diff
@@ -0,0 +1,42 @@
+======
+Target
+======
+First, you must determine whether you want to feature detect on maxima or minima in your dataset. For example, if you are trying to detect clouds in IR satellite data, where clouds have relatively lower brightness temperatures than the background, you would set :code:`target='minimum'`. If, instead, you are trying to detect clouds by cloud water in model data, where an increase in mixing ratio in ...
```
Should feature detect be hyphenated in this usage? If so, can you change here and in other locations where "feature detect" is used as a verb.

pjmarinescu commented on doc/threshold_detection_parameters.rst:
```diff
@@ -0,0 +1,42 @@
+================
+Feature Position
+================
+There are four ways of calculating the single point used to represent feature center: arithmetic center, extreme point, difference weighting, and absolute weighting. Generally, difference weighting (:code:`position_threshold='weighted_diff'`) or absolute weighting (:code:`position_threshold='weighted_abs'`) is suggested for most atmospheric applications. An example of these four methods is shown  ...
```
I was looking at the notebooks, and the titles (i.e., position_threshold='extreme') look like they are associated with the figures above them, even though I believe they represent the code and pictures below them. Therefore, I wonder if you could add more space between the text "position_threshold='XXX' and the figures above them to make this a little more clear.

freemansw1 replied:
I've added lines here between the various plots; hopefully this is enough?

pjmarinescu commented on doc/feature_detection_base_out_vars.csv:
```diff
@@ -0,0 +1,16 @@
+hdim_2,"Second horizontal dimension in grid point space (typically, although not always, E/W or x space)",Number of grid points,float
+num,Number of grid points that are within the threshold of this feature,Number of grid points,int
+threshold_value,Maximum threshold value reached by the feature,Units of the input feature,int
+feature,Unique number of the feature; starts from 1 and increments by 1 to the number of features,n/a,int
```
for the 'feature' variable description, I think it would be more clear if you added at the end of the description "increments by 1 to the number of features **identified in all frames**"

pjmarinescu commented on doc/tracking_output.rst:
```diff
@@ -0,0 +1,11 @@
+Tracking output
+-------------------------
+
+Tracking outputs a `pandas` dataframe with additional variables in addition to the variables output by Feature Detection (see :doc:`feature_detection_output`). The additional variables added by tracking, with column names listed in the `Variable Name` column, are described below with units. 
```
Is the output in the same dataframe as the feature detection output, or a different dataframe? Can you make that more clear?

Also, probably do not need the word "additional" since you have "in addition to" a few words later.

pjmarinescu reviewed (commented):
Hello Sean, 
Overall, I just a few minor comments to make some of the notebooks and variable files a little more clear. Thank you for updating this documentation and adding some new notebooks and variable file lists that will make the use of tobac much easier! - Peter

freemansw1 commented:
Note that I've updated the documentation with the segmentation discussion here. I'm going to wait to request re-reviews until I add in the tracking discussion.

snilsn commented on doc/tracking_base_out_vars.csv:
```diff
@@ -0,0 +1,3 @@
+Variable Name,Description,Units,Type
+cell,Tracked cell number; generally starts from 1. Untracked cell value can be set; but by default is -1.,n/a,int
+time_cell,Time since cell was first detected.,typically minutes,object/python timedelta
```
Isn't that always in minutes, since `transform("min")` is used in `add_cell_time(t)` ?

snilsn reviewed (commented):
Very good work!

I left a few comments, mostly just small things or thoughts.

JuliaKukulies commented on doc/data_input.rst:
```diff
@@ -1,32 +1,19 @@
-  - timestr: the date and time of the feature in string format
-  - latitude, longitude: the central lat/lon of the feature
-  - x,y, etc: these are the central location of the feature in the original dataset coordinates
+For the future development of the next major version of tobac (v2.0), we are moving the basic data structures from Iris cubes to xarray DataArrays for improved computing performance and interoperability with other open-source sorftware packages, including the Pangeo project.
```
include link to Pangeo?

JuliaKukulies commented on doc/data_input.rst:
```diff
@@ -1,32 +1,19 @@
+For information on tracking *output*, see :doc:`tracking_output`. 
 
-Note that in future versions of tobac, it is planned to combine both output data types into a single hierarchical data structure containing both spatial and object information. Additional information about the planned changes can be found in the v2.0-dev project, as well as the tobac roadmap
+Note that in future versions of tobac, it is planned to combine both output data types into a single hierarchical data structure containing both spatial and object information. Additional information about the planned changes can be found in the v2.0-dev branch of the main tobac repository, as well as the tobac roadmap.
```
include links to main repo and roadmap?

JuliaKukulies reviewed (commented):
Great @freemansw1 ! The basic examples are very nice and clear and I do not have much to add. Feel free to consider or ignore my minor suggestions. And yes, last check then with the changes for tracking. 

I know it has not been part of this PR, but I am also wondering if we should update the description for [Example notebooks](https://tobac--150.org.readthedocs.build/en/150/examples.html) with the right zenodo link to the data and a link to the v2.0 tutorials in one go?

JuliaKukulies commented:
Committed the publication page now to this PR. Let me know if you have any comments or requests for changes here!

JuliaKukulies commented:
And sorry, I reverted my first two commits because I still had the old index.rst in my branch. Now everything should be clean and the last three commits contain the additions for the publication page. Feel also free to move it somewhere else if you think it is not the best location to have it after the examples.

freemansw1 commented on doc/feature_detection_output.rst:
```diff
@@ -0,0 +1,11 @@
+Feature Detection Output
+-------------------------
+
+Feature detection outputs a `pandas` dataframe with several variables. The variables, (with column names listed in the `Variable Name` column), are described below with units. Note that while these variables come initially from the feature detection step, segmentation and tracking also share some of these variables. See :doc:`tracking_output` for the additional columns added by tracking.
```
Segmentation does not, but you're right that tracking does.

freemansw1 commented on doc/feature_detection/notebooks/multiple_thresholds_example.ipynb:
```diff
@@ -0,0 +1,281 @@
+    "plt.colorbar()\n",
+    "\n",
+    "# Plot all features detected\n",
+    "plt.scatter(x=single_threshold_features['hdim_2'].values, y=single_threshold_features['hdim_1'].values, color='r', label=\"Detected Features\")\n",
```
I've included this as a second plot

freemansw1 commented:
I believe that I've addressed all the comments here.

I haven't yet finished (or even really started) the documentation on segmentation or tracking. I do want to do that, but I'm inclined to do that as a separate PR off of this one, once this one is merged? That will keep the review process simple. Let me know what you think @snilsn @JuliaKukulies @pjmarinescu

JuliaKukulies commented:
Looks good @freemansw1, I am happy with the changes! And I agree that you it would be easier to include the documentation on segmentation and tracking in a separate PR as this one is already quite big.

snilsn commented:
Good work @freemansw1 , I'm happy with these edits and also with splitting the work on the documentation into two PRs.

freemansw1 commented:
Thanks @snilsn and @JuliaKukulies . Given that @pjmarinescu has comments, I'll wait to hear from him before merging.

pjmarinescu reviewed (approved):
Thanks for considering all my comments Sean. Nice job.

freemansw1 commented:
Thanks everyone!
